### DISCUSSION TITLE
Allow using(idisposable) to be used as a boolean in an expression

### DISCUSSION THREAD

**Discussion by wiz0u** [dotnet/csharplang#4643] [Language Ideas]
I feel like it would be nice to be allowed to write statement like:
```
if (using(stream))
{
  stream.CopyTo(destination);
}
else
  Console.WriteLine("stream was null");
```

What this would mean, is that `using` is considered as an expression returning true if stream is non-null (i.e. an IDisposable), and false if stream is null.
It also mean that the IDisposable gets dispose at the end of the full statement block (if-else), or at the end of the current statement if that expression is not part of a control-flow statement.
(of course, if stream is null, there is nothing to dispose, just like using(nullValue) is already accepted now and does nothing)

What do you think?

**Comment by yaakov-h** [comment]
Other than your particular desire to dispose of the object after the last `else`, the following is already legal C# and achieves pretty much the same thing:

```csharp
if (stream is not null) using (stream)
{
  stream.CopyTo(destination);
}
else
  Console.WriteLine("stream was null");
```

**Reply by wiz0u** [reply]
true, but the automatic indentation gets weird because they are considered as two separate statements.

also it's not a particular desire to dispose after the last else, it's just coherent so you can also write `if (!using(stream))`

you can consider my proposal as syntactic sugar

**Reply by yaakov-h** [reply]
automatic formatting is much cheaper to change than the language itself

**Comment by huoyaoyuan** [comment]
Allowing more things in `if` downgrades readability very much. An explicit null check is always understandable.

**Reply by CyrusNajmabadi** [reply]
i tend to agree with this.  Every time i see something going on with `if` statements, i find it to be a but sort of eye-sort.  About the only things i've been *marginally* ok with are allowing `if !(...expr...)`  instead of `if (!(...expr))`.  And:

```c#
if (someNullableBool)
{
}
else
{
}
null
{
}
```

And even those are things i don't feel like there really is enough of a need to do.

**Reply by theunrepentantgeek** [reply]
I agree - _shorter_ isn't automatically _better_.
